src/Widget/Twig/TwigPreprocessor.php

```diff
@@ -118,8 +118,10 @@ public function preprocessEventList(array $events, $preferredLanguage, array $se
                 $query['cdbid'] = $preprocessedEvent['id'];
                 $url->setQuery($query);
             }
-
-            $preprocessedEvent['detail_link'] = $url->__toString();
+            $referer = parse_url($this->request->server->get('HTTP_REFERER'), PHP_URL_HOST);
+            $utmParams = (strpos($url->__toString(), '?')) ? '&' : '?';
+            $utmParams .= 'utm_source=' . urlencode($referer) . '&utm_medium=referral&utm_campaign=widgets';
+            $preprocessedEvent['detail_link'] = $url->__toString() . $utmParams;
```

Comment:
The logic in lines 122-124 can probably be simplified a lot by using `$url->getQuery()` as is done at line 117. I'm not familiar with Guzzle's value objects myself, but I'm pretty sure they have some convenience methods to make this a lot easier to read. Maybe you can pair on it with @konrad126 or @jdrieghe to clean this up a bit.

Reply:
@bertramakers thx for the feedback. 
I just need to check if the url already contains query params (it doesn't matter which params) but the utmParams always need to be added.
On line 122, I could indeed do something like:
`$utmParams = (empty($url->getQuery())) ? '?' : '&';`

Reply:
That would already be an improvement, but what if the URL has a `#` at the end?
For example `test.com/?foo=bar#var` would become `test.com/?foo=bar#var&utm_source=...`

You can avoid cases like this by using `setQuery` on the `$url` variable instead of concatenating.

I THINK this could work:
```
$referer = parse_url($this->request->server->get('HTTP_REFERER'), PHP_URL_HOST); 

// Get a QueryString object and convert it to an array
$query = $url->getQuery()->toArray();

// Add the utm parameters
$query['utm_source'] = urlencode($referer);
$query['utm_medium'] = 'referral';
$query['utm_campaign'] = 'widgets';

// Update the query string on the url
// Apparently this doesn't HAVE to be a `QueryString` object but can also be an array
$url->setQuery($query);

$preprocessedEvent['detail_link'] = $url->__toString();
```

But I didn't test this, it's just pseudo-code based on the methods I could find in PHPStorm when looking at the code.


```

Reply:
Oooh yes! Now I see.
Great feedback!
I've implemented the changes.
IMO, we don't need the `toArray` method

Reply:
Ah you're right, `QueryString` implements `ArrayAccess` so you don't need to convert it to an array first idd.